[clang][Sema] Propagate qualifiers during derived-to-base conversion

antoniofrighetto opened the pull request:
When accessing a field member through a derived-to-base conversion, ensure the qualifiers are propagated to the base type.

Fixes: https://github.com/llvm/llvm-project/issues/127683.

efriedma-quic commented:
Should the result type of the ImplicitCastExpr be volatile in the AST?

AaronBallman commented on clang/test/SemaCXX/derived-to-base-volatile-qual.cpp:
```diff
@@ -0,0 +1,23 @@
+// CHECK-NEXT:     `-BinaryOperator {{.*}} 'volatile int' lvalue '='
+// CHECK-NEXT:       |-MemberExpr {{.*}} 'volatile int' lvalue .Val
+// CHECK-NEXT:       | `-ImplicitCastExpr {{.*}} 'volatile PR127683::Base' lvalue <UncheckedDerivedToBase (Base)>
+void test_volatile_store() {
+  Obj.Val = 0;
+}
```
I'd appreciate a test for volatile read as well as volatile store.

AaronBallman commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -3107,8 +3107,11 @@ Sema::PerformObjectMemberConversion(Expr *From,
 
-  return ImpCastExprToType(From, DestType, CK_UncheckedDerivedToBase,
-                           VK, &BasePath);
+  if (FromType.isVolatileQualified())
+    DestType.addVolatile();
```
Should we be doing this for all qualifiers? `const` seems like it should almost certainly behave the same way, but I think it's a more general property of qualifiers that the derived to base cast would keep the qualifiers.

antoniofrighetto replied:
From intro.object:
> Objects can contain other objects, called subobjects. A subobject can be a member subobject a base class subobject, or an array element.

So, given:
```cpp
struct Base { int X; };
struct Derived : Base {};

const Derived Obj;
```
Seems outright logical that the entirety of Obj must be const. Thought this applied only for cv-qualifiers, but should indeed make sense for all qualifiers; updated, thanks!

antoniofrighetto replied:
EDIT: Moving from:
```cpp
DestType = Context.getQualifiedType(DestType, FromType.getQualifiers());
```
To CVR ones for now:
```cpp
DestType.withCVRQualifiers(FromType.getCVRQualifiers());
```
Per CI failure [here](https://buildkite.com/llvm-project/github-pull-requests/builds/149797#01952e46-e199-450d-978b-2f763cd9b9c0), due to incompatibility between address spaces. Not completely sure if we should setAddressSpace of DestType AS one to the one of FromType, before getQualifiedType.

AaronBallman commented on clang/test/SemaCXX/derived-to-base-volatile-qual.cpp:
```diff
@@ -0,0 +1,23 @@
+// RUN: %clang_cc1 -std=c++20 -fsyntax-only -ast-dump %s | FileCheck %s
```
Should we also have a CodeGen test showing that we do a `store/load volatile`?

antoniofrighetto replied:
Added a test in CodeGen too, thanks!

AaronBallman reviewed (commented):
Thank you for working on this! Please be sure to add a release note to `clang/docs/ReleaseNotes.rst` so users know about the fix.

shafik commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -3107,8 +3107,12 @@ Sema::PerformObjectMemberConversion(Expr *From,
 
-  return ImpCastExprToType(From, DestType, CK_UncheckedDerivedToBase,
-                           VK, &BasePath);
+  Qualifiers FromTypeQuals = FromType.getQualifiers();
```
Maybe add a quote from https://eel.is/c++draft/basic.type.qualifier#def:object,volatile

antoniofrighetto replied:
Added, thanks!

antoniofrighetto commented on clang/test/SemaCXX/derived-to-base-cv-qualifiers.cpp:
```diff
@@ -0,0 +1,34 @@
+// RUN: %clang_cc1 -std=c++20 -fsyntax-only -ast-dump %s | FileCheck %s
+
+// Ensure volatile is preserved during derived-to-base conversion. 
```
Added, thanks! `__ptrauth` is in the process of being upstreamed, whereas `_Atomic` seems to have its own semantics (not part of Qualifiers, but rather QualifiersAndAtomic), and I'm not completely sure it would fit here (also accessing members of an atomic struct is UB). Concerning `__restrict`, shouldn't it be property of the pointer, not the object's type? I tried something like:
```cpp
void test_restrict(Derived *__restrict D) {
  Base *B = D;
}
```
Yet, in DerivedToBase, `__restrict` seems to be not included in Derived qualifiers (so contextually DestType has not it either, and in any case it should be orthogonal?). Rebased to main too.

AaronBallman reviewed (approved):
LGTM!

antoniofrighetto committed:
[clang][Sema] Propagate qualifiers during derived-to-base conversion

When accessing a field member through a derived-to-base conversion,
ensure qualifiers are propagated to the base class subobject.

Fixes: https://github.com/llvm/llvm-project/issues/127683.
